blockchain/v2/demuxer.go

```diff
@@ -0,0 +1,164 @@
+// nolint:unused
+package v2
+
+import (
+	"fmt"
+	"sync/atomic"
+
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+type scFull struct{}
+type pcFull struct{}
+
+type demuxer struct {
+	input     chan Event
+	scheduler *Routine
+	processor *Routine
+	fin       chan error
+	stopped   chan struct{}
+	rdy       chan struct{}
+	running   *uint32
```

Comment:
have you seen the `common/BaseService` struct https://godoc.org/github.com/tendermint/tendermint/libs/common#BaseService ? feels like you're doing exactly the same

Reply:
I also thought routines could implement the service interface. I spent a bit of time trying trying to make it work but it felt wrong for a few reasons.  

Routines expose methods associated with the lifecycle of a finite state machine while services don't. Implementing OnStart and OnReset for FSM didn't really make sense to me as we don't embed the state in the struct itself but instead close over it. This restricts how state in routines is managed but I would suggest this limitation is a good thing. We want
one routine to manage one state and the transition between states to happen from a serialized stream of events. 

Routines will ideally `terminate` of their own accord and need to communicate their final event. This is an important requirement for the blockchain reactor as it needs to communicate when to switch to consensus. 

While it would most likely be possible to modify the routine to adhere to the Service interface, i'm not sure it's worth it since I cannot envision any cases in which we would want either a Service or a Routine and not care about which.